### ISSUE TITLE
Custom EventMonitor and Interceptor are not called 

### ISSUE THREAD

**Issue by ahmadmssm** [Alamofire/Alamofire#3124]
I have created a new Interceptor and added it as per the Advanced usage documentation and I can see that it is added correctly, but when I added a breakpoint I found out that it is not entering this function 

```
public func adapt(_ urlRequest: URLRequest,
               for session: Session,
               completion: @escaping (Result<URLRequest, Error>) -> Void) {
}
```

In case of the custom Interceptor.

and

```
 public func requestDidResume(_ request: Request) {
        loadingIndicatorService.showLoader()
    }
    
    public func request<Value>(_ request: DataRequest,
                               didParseResponse response: AFDataResponse<Value>) {

    }
    
    public func requestDidCancel(_ request: Request) {

    }
```
 In case of the custom EventMonitor.

**Comment by jshier** [comment]
It would be informative if you could provide an example of how you're initializing your `Session` with your `RequestInterceptor` and `EventMonitor`, especially a demo project that we could use to investigate. Also, are you using your own `Result` type or a library that provides one? I've found that can interfere with `adapt` being called.

**Comment by ahmadmssm** [comment]
@jshier Thanks for your reply,
Here's how i intialize Sessio
```
 open lazy var alamofireSession: Session = {
        return Session(configuration: getAlamofireSessionConfiguration(),
                       interceptor: getRequestInterceptor(),
                       eventMonitors: getEventMonitorsList())
    }()

 open func getAlamofireSessionConfiguration() -> URLSessionConfiguration {
        let configuration = URLSessionConfiguration.af.default
        configuration.timeoutIntervalForRequest = 60
        configuration.timeoutIntervalForResource = 60
        configuration.requestCachePolicy = .useProtocolCachePolicy
        configuration.allowsCellularAccess = true
        return configuration
    }

open func getRequestInterceptor() -> RequestInterceptor? {
        return RequestInterceptor(sessionService: sessionService,
                                  refreTokenAPI: getRefreshTokenAPI(),
                                  additionalHeaders: getAdditionalHeaders())
    }
    
 open func getEventMonitorsList() -> [EventMonitor] {
        var eventMonitors: [EventMonitor] = []
        if (self.shouldEnableConsoleLogging()) {
            eventMonitors.append(AlamofireLogger())
        }
        if (self.shouldEnableLoadingIndicator() && loadingIndicatorService != nil) {
            eventMonitors.append(HttpRequestLoadingIndicator(loadingIndicatorService!))
        }
        return eventMonitors
    }
```

I'm using `Codable` for type parsing

**Comment by cnoon** [comment]
Any chance you could provide a sample project? Perhaps through email if you can't post here? Everything looks okay that you've posted so far assuming you're using the `alamofireSession` you are lazily creating to execute the request.

**Comment by ahmadmssm** [comment]
I think it is not easy because the project size is very big.
Is there a problem with creating a lazy executer for `alamofireSession` ?

**Comment by ahmadmssm** [comment]
@cnoon @jshier Any ideas about the issue.

**Comment by jshier** [comment]
Everything should be working and we have the tests to prove it, so this is very odd. You could try using our `Adapter` type to implement your `adapt` logic, and `ClosureEventMonitor` should let you hook into most events (not the generic `didParseResponse` however). If those don't work either then more investigation will be needed.

**Comment by jshier** [comment]
Another avenue of investigation would be to add break points to the default implementations of the `RequestInterceptor` and `EventMonitor` methods you're trying to implement to see if they're called instead of your implementing methods. That would at least confirm that Alamofire is properly calling them internally, they just aren't getting dispatched to the right instance.

**Comment by ahmadmssm** [comment]
I already did that but it is not called @jshier

**Comment by jshier** [comment]
Where did you put the breakpoints?

**Comment by ahmadmssm** [comment]
```
public func adapt(_ urlRequest: URLRequest,
               for session: Session,
               completion: @escaping (Result<URLRequest, Error>) -> Void) {
        if (isFirstTime) {
            isFirstTime = false
            baseURL = urlRequest.url?.baseURL?.absoluteString
        }
        completion(.success(getModifiedURLRequest(urlRequest, session)))
    }
```

```
 public func requestIsRetrying(_ request: Request) {
        let httpRequest = request.request
        logger.debugPrint("⚡️ Retrying request " + httpRequest.debugDescription)
    }
    
    public func request<Value>(_ request: DataRequest,
                               didParseResponse response: AFDataResponse<Value>) {
        logger.debugPrint("HTTP Endpoint: \(response.debugDescription)")
    }
```

**Comment by jshier** [comment]
Yes, as I said, putting them on the default implementations of those methods may give insight into whether they're being called at all. By that I mean at [line 97 in RequestInterceptor.swift](https://github.com/Alamofire/Alamofire/blob/master/Source/RequestInterceptor.swift#L97) and on the [various events in `EventMonitor`](https://github.com/Alamofire/Alamofire/blob/master/Source/EventMonitor.swift#L256) you're trying to implement.

**Comment by ahmadmssm** [comment]
I just tested, it is not called too (line 97 in RequestInterceptor.swift).

**Comment by jshier** [comment]
What are the full implementations of your `RequestInterceptor` and `AlamofireLogger` types?

**Comment by ahmadmssm** [comment]
```
 open lazy var alamofireSession: Session = {
        return Session(configuration: getAlamofireSessionConfiguration(),
                       interceptor: getRequestInterceptor(),
                       eventMonitors: getEventMonitorsList())
    }()

 open func getAlamofireSessionConfiguration() -> URLSessionConfiguration {
        let configuration = URLSessionConfiguration.af.default
        configuration.timeoutIntervalForRequest = 60
        configuration.timeoutIntervalForResource = 60
        configuration.requestCachePolicy = .useProtocolCachePolicy
        configuration.allowsCellularAccess = true
        return configuration
    }

open func getRequestInterceptor() -> RequestInterceptor? {
        return RequestInterceptor(sessionService: sessionService,
                                  refreTokenAPI: getRefreshTokenAPI(),
                                  additionalHeaders: getAdditionalHeaders())
    }
    
 open func getEventMonitorsList() -> [EventMonitor] {
        var eventMonitors: [EventMonitor] = []
        if (self.shouldEnableConsoleLogging()) {
            eventMonitors.append(AlamofireLogger())
        }
        if (self.shouldEnableLoadingIndicator() && loadingIndicatorService != nil) {
            eventMonitors.append(HttpRequestLoadingIndicator(loadingIndicatorService!))
        }
        return eventMonitors
    }
```

**Comment by jshier** [comment]
I'm sorry, I mean the full implementation of the wrapping class, as least as far as it's relevant to the Alamofire usage. These snippets are missing a lot of context.

**Comment by ahmadmssm** [comment]
```
import Alamofire

open class RequestInterceptor: Alamofire.RequestInterceptor {

    private let lock: NSLock
    private var sessionService: SessionService?
    private var refreTokenAPI: BaseRefreTokenAPI?
    private let requestIntercepters: [RequestIntercepter]
    private let additionalHeaders: [String : String]
    private var requestsToRetry: [(RetryResult) -> Void] = []
    private var isFirstTime = true
    public var isRefreshing = false
    public var baseURL: String?
    
    
    init(sessionService: SessionService? = nil,
         refreTokenAPI: BaseRefreTokenAPI? = nil,
         requestIntercepters: [RequestIntercepter] = [],
         additionalHeaders: [String : String]) {
        self.refreTokenAPI = refreTokenAPI
        self.sessionService = sessionService
        self.requestIntercepters = requestIntercepters
        self.additionalHeaders = additionalHeaders
        self.lock = NSLock()
    }

    public func adapt(_ urlRequest: URLRequest,
               for session: Session,
               completion: @escaping (Result<URLRequest, Error>) -> Void) {
        if (isFirstTime) {
            isFirstTime = false
            baseURL = urlRequest.url?.baseURL?.absoluteString
        }
        completion(.success(getModifiedURLRequest(urlRequest, session)))
    }
    
    open func getModifiedURLRequest(_ urlRequest: URLRequest, _ session: Session) -> URLRequest {
        var request = urlRequest
        appendAdditionalHeaders(&request)
        appendAdditionalIntercepters(&request, session)
        configureRefreshToken(&request)
        return request
    }
    
    open func appendAdditionalHeaders(_ urlRequest: inout URLRequest)  {
        if (!additionalHeaders.isEmpty) {
            additionalHeaders.forEach { (arg0, arg1) in
                urlRequest.setValue(arg0, forHTTPHeaderField: arg1)
            }
        }
    }
    
    open func appendAdditionalIntercepters(_ urlRequest: inout URLRequest, _ session: Session) {
        var request = urlRequest
        if (!requestIntercepters.isEmpty) {
            requestIntercepters.forEach { modifiedRequest in
                request = modifiedRequest.getUpdatedURLRequest(request, for: session)
            }
        }
        urlRequest = request
    }
    
    open func configureRefreshToken(_ urlRequest: inout URLRequest)  {
        if (sessionService?.hasValidSession() ?? false) {
            urlRequest.headers.add(.authorization(bearerToken: sessionService!.getAccessToken()))
        }
    }
    
    public func retry(_ request: Request,
                      for session: Session,
                      dueTo error: Error,
                      completion: @escaping (RetryResult) -> Void) {
        if (sessionService?.shoudRetryWithAuthentication() ?? false) {
            lock.lock() ; defer { lock.unlock() }
            if let response = request.task?.response as? HTTPURLResponse {
                if (tokenShouldRefresh(previousResponse: response)) {
                    requestsToRetry.append(completion)
//                    refreshToken(completion: completion)
                }
                else {
                    /// The request did fail to satisfy the refresh condition.
                    /// Return the original error and don't retry the request.
                    completion(.doNotRetryWithError(error))
                }
            }
            /// The request did fail to satisfy the refresh condition.
            /// Return the original error and don't retry the request.
            completion(.doNotRetryWithError(error))
        }
    }
    
    open func tokenShouldRefresh(previousResponse: HTTPURLResponse) -> Bool {
        /// The request did fail due to a 401 Unauthorized response.
        return previousResponse.statusCode == 401
    }
}
```

**Comment by jshier** [comment]
Thanks. Your code in my sample project works just fine, the adapter and a basic `EventMonitor` are hit correctly.

**Comment by ahmadmssm** [comment]
seems strange, so here is my alamofire configurat.
```
open class AlamofireRestClient: BaseHTTPClient {

    private var sessionService: SessionService?
    private var loadingIndicatorService: LoadingIndicatorService?
    public private(set) lazy var networkReachabilityManager: NetworkReachabilityManager? = {
        return NetworkReachabilityManager()
    }()
//    open lazy var alamofireSession: Session = {
//        return Session(configuration: getAlamofireSessionConfiguration(),
//                       interceptor: getRequestInterceptor(),
//                       eventMonitors: getEventMonitorsList())
//    }()
        
    public init(_ sessionService: SessionService? = nil,
                _ loadingIndicatorService: LoadingIndicatorService? = nil) {
        self.sessionService = sessionService
        self.loadingIndicatorService = loadingIndicatorService
        super.init()
    }
    
    open func shouldEnablePlugins() -> Bool { return true }
       
    open func shouldEnableConsoleLogging() -> Bool { return true }

    open func shouldEnableLoadingIndicator() -> Bool { return false }

    open override func shouldEnableRefreshToken() -> Bool {
        return sessionService?.shoudRetryWithAuthentication() ?? false
    }
    
    open func getRefreshTokenAPI() -> BaseRefreTokenAPI? {
        if (shouldEnableRefreshToken()) {
            fatalError("`getRefreshTokenAPI()` function should be implemented.")
        }
        return nil
    }
    
    open func getJSONDecoder() -> JSONDecoder {
        let decoder: JSONDecoder = JSONDecoder()
        decoder.keyDecodingStrategy = getJSONDecoderDecodingStrategy()
        return decoder
    }
    
    open func getJSONDecoderDecodingStrategy() -> JSONDecoder.KeyDecodingStrategy {
        return JSONDecoder.KeyDecodingStrategy.convertFromSnakeCase
    }
    
    open func getRequestInterceptor() -> RequestInterceptor? {
        return RequestInterceptor(sessionService: sessionService,
                                  refreTokenAPI: getRefreshTokenAPI(),
                                  additionalHeaders: getAdditionalHeaders())
    }
    
    open func getAdditionalHeaders() -> [String : String] {
        return [:]
    }
    
    open func getEventMonitorsList() -> [EventMonitor] {
        var eventMonitors: [EventMonitor] = []
        if (self.shouldEnableConsoleLogging()) {
            eventMonitors.append(AlamofireLogger())
        }
        if (self.shouldEnableLoadingIndicator() && loadingIndicatorService != nil) {
            eventMonitors.append(HttpRequestLoadingIndicator(loadingIndicatorService!))
        }
        return eventMonitors
    }
    
    open func getAlamofireSessionConfiguration() -> URLSessionConfiguration {
        let configuration = URLSessionConfiguration.af.default
        configuration.timeoutIntervalForRequest = 60
        configuration.timeoutIntervalForResource = 60
        configuration.requestCachePolicy = .useProtocolCachePolicy
        configuration.allowsCellularAccess = true
        return configuration
    }
    
    open func getAlamofireSession() -> Session {
        return Session(configuration: getAlamofireSessionConfiguration(),
                       interceptor: getRequestInterceptor(),
                       eventMonitors: getEventMonitorsList())
    }
    
    open func getRequestBuilder<API: AlamofireAPI>(api: API) -> DataRequest {
        return getAlamofireSession().getRequestWith(baseURL: getBaseURL(), api: api)
    }

    open func getMultipartRequestBuilder<API: AlamofireAPI>(api: API) -> UploadRequest {
        return getAlamofireSession().getUploadRequestWith(baseURL: getBaseURL(), api: api)
    }

    // https://stackoverflow.com/questions/41136560/setting-alamofire-custom-destination-file-name-instead-of-using-suggesteddownloa
    open func getDownloadRequestBuilder<API: AlamofireAPI>(api: API,
                                                           downloadDestination: @escaping DownloadRequest.Destination) -> DownloadRequest {
        return getAlamofireSession().getDownloadRequestWith(baseURL: getBaseURL(),
                                                       api: api,
                                                       downloadDestination: downloadDestination)
    }
        
    open func isConnectedToInternet() -> Bool {
        return networkReachabilityManager?.isReachable ?? false
    }

    open func cancelSpecificRequest(byEndpoint endpoint: String) {
        getAlamofireSession().session.getAllTasks{ sessionTasks in
            for task in sessionTasks {
                if task.originalRequest?.url?.absoluteString.contains(endpoint) ?? false {
                    task.cancel()
                }
            }
        }
    }
    
    open func canceUploadRequests() {
        getAlamofireSession().session
            .getTasksWithCompletionHandler({ _, uploadTasks, _ in
                uploadTasks.forEach { $0.cancel() }
        })
    }
   
    open func canceDownloadRequests() {
           getAlamofireSession().session
               .getTasksWithCompletionHandler({ _, _, downloadTasks in
                   downloadTasks.forEach { $0.cancel() }
           })
       }
    
    open func cancelAllRequests() {
        getAlamofireSession().session.invalidateAndCancel()
    }
}
```

**Comment by jshier** [comment]
And I assume that class is used as a singleton or a single instance is injected to use sites?

**Comment by ahmadmssm** [comment]
Yes,  I Inject it using a DI Framework.

**Comment by jshier** [comment]
Your call to `getAlamofireSession` will create a new `Session` every time it's called, which is an anti-pattern. Your single instance in the `lazy var` was correct, you should be using a single `Session` for all requests. Are you able to make successful requests?

**Comment by ahmadmssm** [comment]
Actually I was doing this
```
    open lazy var alamofireSession: Session = {
        return Session(configuration: getAlamofireSessionConfiguration(),
                       interceptor: getRequestInterceptor(),
                       eventMonitors: getEventMonitorsList())
    }()
```
I just did that for the debuging purpose only.

All request are working with the wrong implementation.

**Comment by jshier** [comment]
If you try this sample, does it print "didResume" or hit the appropriate breakpoints? 

```swift
final class API {
    static let shared = API()
    
    lazy var session = Session(interceptor: interceptor, eventMonitors: eventMonitors)
    
    var interceptor: RequestInterceptor {
        let adapter = Adapter { (urlRequest, _, completion) in
            var request = urlRequest
            request.headers.add(.authorization(bearerToken: "token"))
            completion(.success(request))
        }
        let retrier = Retrier { (_, _, _, completion) in
            completion(.doNotRetry)
        }
        
        return Interceptor(adapter: adapter, retrier: retrier)
    }
    
    var eventMonitors: [EventMonitor] {
        let monitor = ClosureEventMonitor()
        monitor.requestDidResume = { _ in print("didResume") }
        
        return [monitor]
    }
}
```

**Comment by ahmadmssm** [comment]
Yes it hits the break points.

**Comment by jshier** [comment]
I'm afraid there's not much more we can try here. I suggest you investigate further and attempt to narrow down what could be going wrong. If you do end up finding an actionable Alamofire bug, feel free to open another issue.

**Comment by ahmadmssm** [comment]
OK, thanks

**Comment by ahmadmssm** [comment]
@jshier Just wanted to say thanks, i was able to solve it as a result of the debugging you steps you did with me.

**Comment by jshier** [comment]
That's great! Can you summarize the issue for future reference?

**Comment by ahmadmssm** [comment]
@jshier Yes, The one who used my rest-client class didn't know that we can invoke request by 2 ways, either by calling `session.response().request { }` or `AF.response().request { }`, he used the second one so the session configurations were not added at all, that is why this issue happened.
